hibernate-core/src/main/java/org/hibernate/persister/entity/mutation/UpdateCoordinatorStandard.java

```diff
@@ -850,11 +853,31 @@ private void decomposeAttributeForUpdate(
 			final AttributeAnalysis attributeAnalysisRef = valuesAnalysis.attributeAnalyses.get( attributeIndex );
 			if ( !attributeAnalysisRef.isSkipped() ) {
 				final IncludedAttributeAnalysis attributeAnalysis = (IncludedAttributeAnalysis) attributeAnalysisRef;
-
 				if ( attributeAnalysis.includeInSet() ) {
 					// apply the new values
-					if ( includeInSet( dirtinessChecker, attributeIndex, attributeMapping, attributeAnalysis ) ) {
-						decomposeAttributeMapping(session, jdbcValueBindings, tableMapping, attributeMapping, values[attributeIndex] );
+					if ( attributeMapping.isEmbeddedAttributeMapping() ) {
+						decomposeEmbeddedAttributeMapping(
+								values[attributeIndex],
+								session,
+								jdbcValueBindings,
+								tableMapping,
+								entityPersister().getEntityMetamodel().isDynamicUpdate() && dirtinessChecker != null,
+								dirtinessChecker != null ?
+										dirtinessChecker.isDirty( attributeIndex, attributeMapping ).isDirty() :
```

Comment:
I'm a bit concerned that dirtiness checking is so high level i.e. just for root attributes. Shouldn't this be part of the `AttributeAnalysis`? Also, can that maybe check nested dirtiness automatically when using dynamic updates? I think that would make the whole process a bit simpler as we could pass along the `AttributeAnalysis` info into the decomposition.

Reply:
My idea is that while building `UpdateValuesAnalysisImpl` and the respective `AttributeAnalysis`, we could compute e.g. a bit set that says if a selectable index is to be included or not. That way, we could have a very simple check in the decompose step and push the logic with generators into the analysis phase where it IMO belongs. We can even pre-compute some model based info to only do checks at runtime that really rely on the value/dirtiness.